src/sandstorm/util/http.c++

```diff
@@ -0,0 +1,35 @@
+#include "http.h"
+
+namespace sandstorm::util::http {
+
+ExtraHeadersResponse::ExtraHeadersResponse(kj::HttpService::Response& orig, kj::HttpHeaders&& extraHeaders)
+  : origResponse(orig),
+  extraHeaders(kj::mv(extraHeaders)) {}
+
+void ExtraHeadersResponse::updateHeaders(const kj::HttpHeaders& headers) {
+  // Update `extraHeaders` by copying the contents of `headers` into it.
+  kj::HttpHeaders newHeaders = headers.cloneShallow();
+  extraHeaders.forEach(
+    [&](kj::HttpHeaderId id, kj::StringPtr value) {
+      newHeaders.set(id, value);
+    },
+    [&](kj::StringPtr name, kj::StringPtr value) {
+      newHeaders.add(name, value);
+    }
+  );
+  extraHeaders = kj::mv(newHeaders);
```

Comment:
It's not necessary to store the new headers as a class member, because the contract of `Response::send()` states that the headers can be destroyed as soon as the method returns. So you could put the new object on the stack instead.

I guess it doesn't actually make a difference in the end, but feels kind of weird to modify the class state like this.

Reply:
Aha, I'd missed that; that does make this much simpler.